File: Bin/records.pl
#!/opt/perl/bin/perl

use 5.032;

use strict;
use warnings;
no  warnings 'syntax';

use experimental 'signatures';
use experimental 'lexical_subs';

use Getopt::Long;
use Date::Calc qw [Add_Delta_Days Delta_Days leap_year Day_of_Year];
use JSON;

sub process_file;
sub sec2time;
sub date2value;

my $distances = {
    big       => [qw [500m 5000m 1500m 10000m]],
    small     => [qw [500m 3000m 1500m  5000m]],
    old_small => [qw [500m 3000m 1000m  5000m]],
    mini      => [qw [500m 1500m 1000m  3000m]],
    sprint    => [qw [500m 1000m  500m  1000m]],
   '2x500'    => [qw [500m  500m]],
};

GetOptions ("input-dir=s"  =>  \my $input_dir,
            "output=s"     =>  \my $output,);

die "Usage: $0 --input_dir <dir> --output <file>" unless $input_dir && $output;

if (-f $input_dir && ! -d $input_dir) {
    #
    # We accept a file inside the input_dir. We do this so we can make
    # use of some Makefile variables.
    #
    $input_dir =~ s!/+[^/]+$!!;
}

die "$input_dir is not a directory" unless -d $input_dir;

$output  .= ".js"  unless $output =~ /\.js$/;

open my $o_fh, ">", $output or die "Failed to open $output: $!";
print $o_fh <<~ "--";
Record . init = function () {
--

opendir my $dh, $input_dir or die "Failed to opendir $input_dir: $!";
while (my $file = readdir $dh) {
    next unless $file =~ /\.txt/;
    process_file "$input_dir/$file", $o_fh;
}
print $o_fh <<~ "--";
}
--
close $o_fh  or die "Failed to close $output: $!";
closedir $dh or die "Failed to close $input_dir: $!";

sub process_file ($input, $o_fh) {
    state $genders     = [qw [men women mixed]];
    state $distance    = [qw [500 1000 1500 3000 5000 10000]];
    state $combination = [qw [big small mini old_small sprint 2x500]];
    state $team        = [qw [pursuit team_sprint relay]];
    state $p_skater    = "(?:[-a-z_]+)";
    state $p_country   = "(?:[A-Z]{3})";
    state $p_rink      = "(?:[-a-z_]+)";
    state $p_result    = "(?:[0-9:.]+)";
    state $p_date      = "(?:[0-9]{4}-[0-9]{2}-[0-9]{2})";
    local $" = "|";
    $input =~ m!/(?<gender>@$genders) _
                 (?<discipline>@$distance|@$combination|@$team) \.txt $!x
              or die "Failed to parse $input";
    my $gender         = $+ {gender};
    my $discipline     = $+ {discipline};
    my $result_is_time = $discipline =~ /^(?:\d+|pursuit|team_sprint|relay)$/;

    my $event = {
        sport          => 'speedskating',
        age            => 'senior',
        type           => 'world',
        gender         =>  $gender,
        discipline     =>  $discipline,
        is_combination => scalar (grep {$_ eq $discipline} @$combination),
        is_team        => scalar (grep {$_ eq $discipline} @$team),
    };

    if ($$distances {$discipline}) {
        $$event {distances} = $$distances {$discipline};
    }

    open my $i_fh, "<", $input or die "Failed to open $input: $!";
    local $/ = undef;

    my $text = <$i_fh>;
       $text =~ s/^\h*#.*\n//gm;

    while ($text =~ s/^ \h* %% \h* (?<key>[a-z_]+): \h*
                                   (?<value> (?:\S .* (?:\n\h{4,} .*)*)? \S)
                                             \h*\n//mx) {
        my $key   = $+ {key};
        my $value = $+ {value} =~ s/\n\h{4,}/ /gr;
        if ($key eq "scale_y_min" && $value =~ /([0-9]+):([0-9]+)/) {
            $value = 60 * $1 + $2;
        }

        $$event {$key} = $value;
    }

    $event = encode_json ($event);

    print $o_fh "    Record . event ($event)\n";


    my @entries = split /^(?=[a-z])/im => $text;
    my @output;
    foreach my $entry (@entries) {
        if ($entry =~ /^SUSPENDED \h+ (?<date> $p_date) \h* \n
                        \s* (?<message> (?s:.*)) \n/x) {
            my $date    = $+ {date};
            my $message = $+ {message} =~ s/\s+/ /gr;
            my $new     = {
                date      => $date,
                message   => $message,
                suspended => 1,
                x         => date2value ($date),
            };
            push @output => $new;
            next;
        }
        $entry =~ m{^(?:(?<skater>  $p_skater) |
                        (?<country> $p_country))             \h+
                     (?<result>  $p_result)                  \h+
                     (?<date>    \d{4}-\d{2}-\d{2} (?:/\d)?) \h+
                     (?<rink>    $p_rink)                    \h* \n
                     (?<members> (?:\h+ $p_skater)* \n)?
                     (?<times>   (?:\h+ $p_result)* \n)?
                    }x or die "Cannot parse $entry";

        my $skater  = $+ {skater};
        my $country = $+ {country};
        my $result  = $+ {result};
        my $date    = $+ {date};
        my $rink    = $+ {rink};
        my $times   = $+ {times};
        my $members = $+ {members};
           $times   = [$times   =~ /\S+/g] if $times;
           $members = [$members =~ /\S+/g] if $members;

        my $precision = $result =~ /\.(\d+)$/ ? length ($1) : 0;

        my $new = {
            venue     =>   $rink,
            result    =>   $result,
            precision =>   $precision,
        };

        $$new {athlete} = $skater  if $skater;
        $$new {country} = $country if $country;
        $$new {times}   = $times   if $times;
        $$new {members} = $members if $members;

        if ($result_is_time) {
            $result =~ /^(?:(?:(?<hours>\d+):)?
                               (?<minutes>\d{1,2}):)?
                               (?<seconds>\d{1,2}(?:\.\d{1,3})?)$/x
                or die "Failed to parse result '$result'";
            $$new {time_in_sec} = ($+ {hours}   || 0) * 3600 +
                                  ($+ {minutes} || 0) *   60 +
                                   $+ {seconds};
        }

        $date =~ m     [^(?<year> [0-9]{4}) -
                         (?<month>[0-9]{2}) -
                         (?<day>  [0-9]{2}) 
                     (?:/(?<duration>[1-3]))?]x
                 or die "Failed to parse date '$date'";

        my ($year, $month, $day, $duration) = @+ {qw [year month day duration]};
        my $first_date = "$year-$month-$day";
        my $last_date  =  $first_date;
        if ($duration && $duration > 1) {
            my ($y, $m, $d) = Add_Delta_Days ($year, $month, $day,
                                              $duration - 1);
            $last_date = sprintf "%04d-%02d-%02d" => $y, $m, $d;
        }
        $$new {first_date} = $first_date;
        $$new {date}       = $last_date;
        $$new {year}       = $year;
        $$new {month}      = $month;
        $$new {day}        = $day;

        #
        # Calculate the improvement. This can only be done if:
        #  - The record itself isn't suspended
        #  - The previous record exists
        #  - The previous record isn't suspended
        #
        if (@output) {
            my $previous = $output [-1];
            unless ($$previous {suspended} || $$new {suspended}) {
                my $key = $result_is_time ? "time_in_sec" : "result";
                my $improvement = $$previous {$key} - $$new {$key};
                   $improvement = 0 if $improvement < 0;
                if ($improvement && $improvement > 0) {
                    my $width = $$previous {precision} < $$new {precision}
                              ? $$previous {precision} : $$new {precision};
                    $$new {improvement} = sprintf "%.${width}f" => $improvement;
                    $$new {improvement_formatted} =
                       $result_is_time ? sec2time ($$new {improvement}, $width)
                                       :           $$new {improvement};
                }
            }
        }

        #
        # x and y values are required for the plotting.
        #
        $$new {x} = date2value ($last_date);  # Date as a float.
        $$new {y} = $result_is_time ? $$new {time_in_sec} : $$new {result};

        push @output => $new;
    }

    #
    # Calculate the duration. For each record, find the first record
    # which is an improvement (to deal with equalized records), or
    # a suspension and calculate the duration.
    #
    # "Current" records will *not* get a duration, nor will suspensions.
    #
    for (my $i = 0; $i < @output; $i ++) {
        my $entry = $output [$i];
        next if $$entry {suspended};
        for (my $j = $i + 1; $j < @output; $j ++) {
            my $next = $output [$j];
            next if !$$next {suspended} && !$$next {improvement};
            $$entry {duration} = Delta_Days (split (/-/ => $$entry {date}),
                                             split (/-/ => $$next  {date}));
            last;
        }
    }

    #
    # Find the "current" records. Walk from the end, as long as there is
    # neither an improvement, or a suspension, it's current.
    #
    for (my $i = @output - 1; $i >= 0; $i --) {
        my $entry = $output [$i];
        last if $$entry {suspended} || $$entry {duration};
        $$entry {current} = 1;
    }

    foreach my $entry (@output) {
        my $json = encode_json ($entry);
        my $method = $$entry {suspended} ? "add_suspension" : "add_record";
        printf $o_fh "    . $method ($json)\n";
    }
    print $o_fh "\n";
    close $i_fh or die "Failed to close $input: $!";
}



sub sec2time ($seconds, $precision) {
    my $time = "";
    if ($seconds >= 3600) {
        my $hours = int ($seconds / 3600);
        $seconds -= $hours * 3600;
        $time = "$hours:";
    }
    my $minutes = int ($seconds / 60);
    $seconds -= $minutes * 60;
    $time .= "0" if $time && $minutes < 10;
    $time .= "$minutes:";
    my $width = 3 + $precision;
    $time .= sprintf "%0${width}.${precision}f" => $seconds;
}


sub date2value ($date) {
    my ($year, $month, $day) = split /-/ => $date;
    my $doy  = Day_of_Year ($year, $month, $day) - 1;
    my $days = 365 + (leap_year ($year) ? 1 : 0);
    $year + $doy / $days;
}

File: JavaScript/Sports/Records.js
let genders      = ['men', 'women', 'mixed']
let ages         = ['senior']
let types        = ['world']
let disciplines  = {
    speedskating: ['500', '1000', '1500', '3000', '5000', '10000',
                   'big', 'small', 'old_small', 'mini', 'sprint', '2x500',
                   'pursuit', 'team_sprint', 'relay'],
}
class Record {
    static records = []

    constructor (args) {
        this . __sport          = args . sport
        this . __discipline     = args . discipline
        this . __gender         = args . gender
        this . __type           = args . type
        this . __age            = args . age
        this . __full_name      = args . full_name
        this . __short_name     = args . short_name
        this . __is_combination = args . is_combination
        this . __is_team        = args . is_team
        this . __distances      = args . distances
        this . __progression    = []
        this . __summary        = args . summary || ""
    }

    static genders () {
        return genders
    }

    static disciplines (sport) {
        return disciplines [sport]
    }

    //
    // Return the event with the given parameters, or create a
    // new one, and return that one.
    //
    static event (args = {}) {
        //
        // Defaults
        //
        if (!args . age)  {args . age  = 'senior'}
        if (!args . type) {args . type = 'world'}

        let gender      = args . gender      || ""
        let sport       = args . sport       || ""
        let discipline  = args . discipline  || ""
        let age         = args . age
        let type        = args . type
        let exists      = args . exists

        if (!genders . includes (gender)) {
            console . log (`${gender} is not a gender`)
            return
        }
        if (!ages . includes (age)) {
            console . log (`${age} is not a valid age category`)
            return
        }
        if (!types . includes (type)) {
            console . log (`${type} is not a valid record type`)
            return
        }
        if (!disciplines [sport]) {
            console . log (`${sport} is not a valid sport`)
            return
        }
        if (!disciplines [sport] . includes (discipline)) {
            console . log (`${discipline} is not a valid ` +
                            `discipline for ${sport}`)
            return
        }

        if (!Record . records [sport]) {
            if (exists) {return undefined}
            Record . records [sport] = []
        }
        if (!Record . records [sport] [discipline]) {
            if (exists) {return undefined}
            Record . records [sport] [discipline] = []
        }
        if (!Record . records [sport] [discipline] [gender]) {
            if (exists) {return undefined}
            Record . records [sport] [discipline] [gender] = []
        }
        if (!Record . records [sport] [discipline] [gender] [type]) {
            if (exists) {return undefined}
            Record . records [sport] [discipline] [gender] [type] = []
        }
        if (!Record . records [sport] [discipline] [gender] [type] [age]) {
            if (exists) {return undefined}
            Record . records [sport] [discipline] [gender] [type] [age] = 
                new Record (args)
        }

        return Record . records [sport] [discipline] [gender] [type] [age]
    }

    //
    // Returns the event with the given parameters, or undefined
    // if it doesn't exist
    //
    static exists (args = {}) {
        args . exists = 1
        return this . event (args)
    }

    //
    // Getters
    //
    sport          () {return this . __sport}
    discipline     () {return this . __discipline}
    gender         () {return this . __gender}
    type           () {return this . __type}
    age            () {return this . __age}
    is_combination () {return this . __is_combination}
    is_team        () {return this . __is_team}
    distances      () {return this . __distances}
    progression    () {return this . __progression}
    summary        () {return this . __summary}

    //
    // Return the current record
    //
    current () {
        return this . __progression . filter ((entry) => {
            return !entry . is_suspension () &&
                    entry . is_current    ()
        })
    }

    //
    // Add a record/suspension; it is assumed records are added in order
    //
    add_record (args) {
        let new_entry = new Record_Entry ({...args, record: this})
        this . __progression . push (new_entry)
        return this
    }
    add_suspension (args) {
        let new_entry = new Suspension ({... args, record: this})
        this . __progression . push (new_entry)
        return this
    }

    //
    // (Full) name of the discipline
    //
    name (short = false) {
        return short ? this . __short_name : this . __full_name
    }
}


class Record_Entry {
    constructor (args) {
        this . __record       =                    args . record
        this . __venue        = Venue   . venue   (args . venue)
        this . __result       =                    args . result
        this . __date         =                    args . date
        this . __first_date   =                    args . first_date
        this . __current      = args . current ? true : false
        this . __precision    =                    args . precision
        if (args . improvement) {
            this . __improvement  =                args . improvement
        }
        if (args . improvement_formatted) {
            this . __improvement_formatted =       args . improvement_formatted
        }
        if (args . time_in_sec) {
            this . __time_in_sec  =                args . time_in_sec
        }
        if (args . athlete) {
            this . __athlete  = Athlete . athlete (args . athlete)
        }
        if (args . country) {
            this . __country  = Country . country (args . country)
        }
        if (args . times) {
            this . __times    =                    args . times
        }
        if (args . members) {
            this . __members  =  args . members . map ((ath) =>
                                     Athlete . athlete (ath))
        }

        //
        // Set the duration, or calculate it.
        //
        if (args . duration != null) {
            this . __duration =                    args . duration
        }
        else {
            this . __duration = Math . round (
                    (new Date ()              . setHours (12) -
                     new Date (this . __date) . setHours (12)) /
                    (24 * 3600 * 1000))
        }
    }

    //
    // Getters
    //
    athlete               () {return this . __athlete}
    country               () {return this . __country}
    venue                 () {return this . __venue}
    result                () {return this . __result}
    date                  () {return this . __date}
    first_date            () {return this . __first_date}
    time_in_sec           () {return this . __time_in_sec}
    improvement           () {return this . __improvement}
    improvement_formatted () {return this . __improvement_formatted}
    times                 () {return this . __times || []}
    members               () {return this . __members}
    duration              () {return this . __duration}
    is_current            () {return this . __current}
    precision             () {return this . __precision}
    record                () {return this . __record}

    athlete_or_team       () {return this . __athlete ||
                                     this . __country}

    is_suspension       () {return false}
}



class Suspension {
    constructor (args) {
        this . __date    = args . date
        this . __message = args . message
        this . __record  = args . record
    }

    is_suspension () {return true}

    date    () {return this . __date}
    message () {return this . __message}
}
